### PR TITLE
DBZ-2891 Oracle Xstream deadlock work

### PR DESCRIPTION
**Pull request by Naros**
https://issues.redhat.com/browse/DBZ-2891

@jpechane This PR isn't yet ready, this contains just the integration test that I'm using to try and find a solution.  So far I haven't been able to devise a solution using a type of lock/flag in order to avoid the deadlock reliably.  I'll try to circle back with you when you arrive in the morning to discuss next steps.

### CODE REVIEW

**Comment by Naros** [comment]
@jpechane This is ready for review.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/XstreamStreamingChangeEventSource.java [hunk 1]
```diff
@@ -139,4 +159,23 @@ else if (scn != null) {
...
+        // This is to mimic the behavior of the Xstream callback API
+        if (lcr instanceof RowLCR) {
+            if (((RowLCR) lcr).hasChunkData()) {
+                ChunkColumnValue chunk;
+                do {
+                    chunk = xsOut.receiveChunk(XStreamOut.DEFAULT_MODE);
+                    handler.processChunk(chunk);
+                } while (!chunk.isEndOfRow());
+            }
+        }
```
Comment (Naros) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/XstreamStreamingChangeEventSource.java:179]: We could debate whether this is even necessary given the fact that the callback handler isn't written to support chunks but I opted to include this mainly to remain consistent with what would have be observed by using the callback handler previously if we had encountered a row that had chunk data.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/XstreamStreamingChangeEventSource.java [hunk 2]
```diff
@@ -71,11 +76,26 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
...
+                    // When a batch begins, Xstream expects that to end before a new operation can
+                    // be performed, which includes detachment. This is why we only test for context
+                    // shutdown once we've left the LCR batch execution status.
+                    if (!context.isRunning()) {
+                        break;
+                    }
+
+                    // Xstream callback API implies a 1 second delay between batches
+                    // To mimic that behavior with the non-callback API, we apply a pause of 1 second
+                    metronome.pause();
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/XstreamStreamingChangeEventSource.java:97]: So to be clear, is this something that happens "often"? Or only rarely? I.e. is there any impact on latency in the common case? Or would it have been the same with the previous approach, too?

Comment (Naros) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/XstreamStreamingChangeEventSource.java:97]: So the callback API, e.g. `receiveLCRCallback` internally applies a 1 second delay between batches, and goal was to keep that the same.  This meant that if an LCR was returned, we wouldn't apply the pause until we had drained all the LCR records from Xstreams and we would then apply a 1 second poll until a new record was detected.

**Comment by Naros** [comment]
Closing this PR in favor of #243.  I'll leave this branch around in case we need to resurrect the changes.

**Pull request closed without merging**